### ISSUE TITLE
nginx proxy trouble for media: permission denied while reading upstream

### ISSUE THREAD

**Issue by isosphere** [girlbossceo/conduwuit#505]
TL;DR: my nginx proxy cannot serve media uploaded to conduwuit.

I am running 20aab1792be1b637bee4e268a7bd8d163321fb63 on Debian bookworm; no docker. 

The `conduwuit` binary is at `/usr/local/bin/conduit`. 
The config file is at `/etc/conduwuit.toml`.

Here's my nginx config:

```nginx
upstream matrix {
    server 127.0.0.1:6167;
}

server {
    listen 8448;
    listen 443;

    server_name m.dataheck.com;

    client_max_body_size    10M;
    keepalive_timeout       15;

    access_log  /var/log/nginx/access.matrix.log;
    error_log   /var/log/nginx/error.matrix.log;

    ssl on;
    ssl_certificate /etc/letsencrypt/live/m.dataheck.com/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/m.dataheck.com/privkey.pem;
    ssl_session_cache shared:SSL:1m;
    ssl_session_timeout  10m;
    ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
    ssl_ciphers HIGH:SEED:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!RSAPSK:!aDH:!aECDH:!EDH-DSS-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA:!SRP;
    ssl_prefer_server_ciphers on;

    location /.well-known/matrix/server {
        return 200 '{"m.server": "m.dataheck.com:443"}';
        types { } default_type "application/json; charset=utf-8";
    }

    location /.well-known/matrix/client {
        return 200 '{"m.homeserver": {"base_url": "https://m.dataheck.com"}}';
        types { } default_type "application/json; charset=utf-8";
        add_header "Access-Control-Allow-Origin" *;
    }

    location /.well-known {
        root /home/matrix/public/;
    }

    location ~* ^(\/_matrix|\/_synapse\/client) {
        proxy_pass http://matrix;
        proxy_set_header X-Forwarded-For $remote_addr;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header Host $host;
    }
}
```

Here's my `systemd` service file:

```systemd
[Unit]
Description=conduwuit Matrix homeserver
After=network.target
Documentation=https://conduwuit.puppyirl.gay/
RequiresMountsFor=/var/lib/private/conduwuit

[Service]
DynamicUser=yes
Type=notify

AmbientCapabilities=
CapabilityBoundingSet=

DevicePolicy=closed
LockPersonality=yes
MemoryDenyWriteExecute=yes
NoNewPrivileges=yes
ProcSubset=pid
ProtectClock=yes
ProtectControlGroups=yes
ProtectHome=yes
ProtectHostname=yes
ProtectKernelLogs=yes
ProtectKernelModules=yes
ProtectKernelTunables=yes
ProtectProc=invisible
ProtectSystem=strict
PrivateDevices=yes
PrivateMounts=yes
PrivateTmp=yes
PrivateUsers=yes
PrivateIPC=yes
RemoveIPC=yes
RestrictAddressFamilies=AF_INET AF_INET6 AF_UNIX
RestrictNamespaces=yes
RestrictRealtime=yes
RestrictSUIDSGID=yes
SystemCallArchitectures=native
SystemCallFilter=@system-service @resources
SystemCallFilter=~@clock @debug @module @mount @reboot @swap @cpu-emulation @obsolete @timer @chown @setuid @privileged @keyring @ipc
SystemCallErrorNumber=EPERM
StateDirectory=conduwuit
RuntimeDirectory=conduwuit
RuntimeDirectoryMode=0750

Environment="CONDUWUIT_CONFIG=/etc/conduwuit.toml"
BindPaths=/var/lib/private/conduwuit:/var/lib/conduwuit
BindPaths=/var/lib/private/conduwuit:/var/lib/private/matrix-conduit

ExecStart=/usr/local/bin/conduit
Restart=on-failure
RestartSec=5

TimeoutStopSec=4m
TimeoutStartSec=4m

StartLimitInterval=1m
StartLimitBurst=5

[Install]
WantedBy=multi-user.target
```

I originally used the example Debian systemd file, but I found the arch one was more recently updated and opted to see if that would resolve the issue (it did not).

This nginx error log is representive of the issue that occurs whenever a user attempts to load an image in any client:

```
2024/07/03 20:19:44 [crit] 5488#5488: *3262 open() "/var/lib/nginx/proxy/3/52/0000000523" failed (13: Permission denied) while reading upstream, client: x.x.x.x, server: m.dataheck.com, request: "GET /_matrix/media/r0/download/m.dataheck.com/Ror46zzPm77RvudC80bhW0bwqHvWARBL?allow_remote=false HTTP/1.1", upstream: "http://127.0.0.1:6167/_matrix/media/r0/download/m.dataheck.com/Ror46zzPm77RvudC80bhW0bwqHvWARBL?allow_remote=false", host: "m.dataheck.com:443"
```

If I manually fetch `http://127.0.0.1:6167/_matrix/media/r0/download/m.dataheck.com/Ror46zzPm77RvudC80bhW0bwqHvWARBL?allow_remote=false` with `wget`, it works. I've done this as my regular user, and as the `nginx` user, no trouble - I get an image. The proxy can't do this for media, though. 

I have nuked my database and started again to no avail.

Any assistance is appreciated.

**Comment by Aranjedeath** [comment]
check your location line, mine is more like:
```
  location /_matrix/ {
      add_header 'Access-Control-Allow-Methods' 'GET, POST, PUT, DELETE, OPTIONS, PATCH, HEAD';
      proxy_pass http://[::1]:6167;
      proxy_set_header Host $http_host;
      proxy_set_header X-Forwarded-For $remote_addr;
      proxy_buffering off;
      client_max_body_size 200M;
      proxy_read_timeout 10m;
  }
```

Specifically:
```
location ~* ^(\/_matrix|\/_synapse\/client) {
```
this make me nervous that it is not catching all matrix requests, and so some (media) are not making it to conduwuit.

**Comment by isosphere** [comment]
I did some A/B testing with the differen

**Comment by isosphere** [comment]
I did some A/B testing with the differences between our configs, and I found that the "secret sauce" was this line:

`proxy_buffering off;`

I'll open a draft PR for adding an example nginx config to the `/debian` directory. I'll try to make it a reasonable file (i'll use a more direct location directive). Having a known-good file to start with would have saved me a lot of time.

Thank you for your help!

**Comment by girlbossceo** [comment]
conduwuit intentionally deleted "example setup configs" to refrain from duplicating the tons of guides out there on basic reverse proxy usage, maintenance burden on these files, and people asking "why do you have this but not that?" which leads to more and more "example setup configs". Upstream has interests in doing this too for similar reasons, so we won't accept a "sample nginx config" PR.

A suitable alternative would be something like this: https://conduwuit.puppyirl.gay/deploying/generic.html#setting-up-the-reverse-proxy

A one line statement mentioning a small caveat.
